vllm/transformers_utils/tokenizer.py

```diff
@@ -6,11 +9,46 @@
 from vllm.logger import init_logger
 from vllm.lora.request import LoRARequest
 from vllm.utils import make_async, LRUCache
+from vllm.engine.ray_utils import ray
 from vllm.transformers_utils.tokenizers import *
 
 logger = init_logger(__name__)
 
 
+def _get_cached_tokenizer(
+    tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+) -> Union[PreTrainedTokenizer, PreTrainedTokenizerFast]:
+    """Get tokenizer with cached properties.
+
+    By default, transformers will recompute multiple tokenizer properties
+    each time they are called, leading to a significant slowdown. This
+    function caches these properties for faster access."""
+
+    tokenizer_all_special_ids = set(tokenizer.all_special_ids)
+    tokenizer_all_special_tokens_extended = (
+        tokenizer.all_special_tokens_extended)
+    tokenizer_all_special_tokens = set(tokenizer.all_special_tokens)
+
+    class CachedTokenizer(tokenizer.__class__):
+
+        @property
+        def all_special_ids(self):
+            return tokenizer_all_special_ids
+
+        @property
+        def all_special_tokens(self):
+            return tokenizer_all_special_tokens
+
+        @property
+        def all_special_tokens_extended(self):
+            return tokenizer_all_special_tokens_extended
```

Comment:
This kind of thing is hacky but can be justified if the performance gain is enough.

Assuming the performance gain is worth it, can we add some good tests to defend against changes in the upstream class?

I don't have a full understanding on the differences between `all_special_tokens_extended` and the others; if we have tests that guarantee correctness of this caching class then I don't have to learn the details to understand if the class if functioning correctly in the face of upstream changes.

Reply:
This is critical to get good performance out of HF tokenizers.

Reply:
ack!